[lld] Do not implicitly link non "public" libraries

drodriguez opened the pull request:
The LC_SUB_CLIENT Mach-O command and the `allowable-clients` TBD entry specify that the given framework (or library?) can only be linked directly from the specified names, even if it is sitting in `/usr/lib` or `/System/Library/Frameworks`.

Add a check for those conditions before checking if a library should be implicitly linked, and link against their umbrella if they have allowable clients. The code needs to be in both the binary libraries and the interface libraries.

Add a test that reproduces the scenario in which a framework reexports a private framework that sits in `/System/Library/Frameworks`, and check for the symbols of the reexported framework to be associated with the public framework, and not the private one.

drodriguez committed:
[lld] Do not implicitly link non "public" libraries

The LC_SUB_CLIENT Mach-O command and the `allowable-clients` TBD entry
specify that the given framework (or library?) can only be linked
directly from the specified names, even if it is sitting in `/usr/lib` or
`/System/Library/Frameworks`.

Add a check for those conditions before checking if a library should be
implicitly linked, and link against their umbrella if they have
allowable clients. The code needs to be in both the binary libraries and
the interface libraries.

Add a test that reproduces the scenario in which a framework reexports
a private framework that sits in `/System/Library/Frameworks`, and check
for the symbols of the reexported framework to be associated with the
public framework, and not the private one.

tstellar reviewed (commented):
You have a typo in the PR summary:

s/ldd/lld/

ellishg commented on lld/test/MachO/implicit-and-allowable-clients.test:
```diff
@@ -0,0 +1,49 @@
+
+# RUN: llvm-objdump --bind --no-show-raw-insn -d %t/test | FileCheck %s
+# CHECK:     Bind table:
+# CHECK-DAG: __DATA __data {{.*}} pointer 0 Framework1   _func1
+# CHECK-DAG: __DATA __data {{.*}} pointer 0 Framework1   _func11
```
You might want to avoid using names `_func1` and `_func11` since the first is a substring of the other. Line 10 could match the next line.

drodriguez committed:
[lld][test] Avoid names that can be prefixes

The function and framework names could had matched incorrectly because
ones were prefixes of the others. Use more clear and non-prefix names
for the frameworks and the functions in the test.

dyung commented on lld/test/MachO/implicit-and-allowable-clients.test:
```diff
@@ -0,0 +1,49 @@
+# REQUIRES: arm
```
Shouldn't this be AArch64, not arm?

drodriguez replied:
I thought `arm` implied `aarch64`, doesn't it? I can change it. It is actually not important which architecture is used for building. I just needed one.

drodriguez replied:
Seems that it has already been changed in ce863dd5a77723fc21e1e43f5d3dd503e9a73d14.

I was under the impression that the arm backend implied both of them. I see there's two backends there. I will try not to forget for next time.

dyung replied:
It does not, arm and aarch64 are distinct, but I see @MaskRay has already updated the REQUIRES line in the test.

dyung reviewed (commented):
Was this change approved before it was submitted?

bdash commented:
Apple's linker rejects linking to libraries with allowable client lists:

```
$ cc -framework vecLib -o test test.m         
ld: cannot link directly with 'vecLib' because product being built is not an allowed client of it
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

After this change, not only does lld not reject linking to vecLib.framework, it links to the library despite the binary not being an allowed client:
```
$ cc -framework vecLib -fuse-ld=lld -Bllvm-build/Release+Asserts/bin -o test test.m
$ nm -m test | grep vecLib                                                                                               
                 (undefined) external _vvdiv (from vecLib)
$ grep ' clients:' $(xcrun --show-sdk-path)/System/Library/Frameworks/Accelerate.framework/Versions/Current/Frameworks/vecLib.framework/vecLib.tbd
    clients:         [ Accelerate ]
    clients:         [ vecLib ]
    clients:         [ vecLib ]
    clients:         [ vecLib ]
    clients:         [ vecLib ]
    clients:         [ Sparse, vecLib ]
```

What was the motivation for this change? Generating binaries that don't respect the allowable client list means applications will break if / when Apple moves symbols between libraries. Is this intentional?

There's no bug associated with this change so I cannot tell what problem it is trying to solve.

test.m:
```
#include <stdio.h>
#include <vecLib/vForce.h>

int main(int argc, char**argv) {
    fprintf(stderr, "%p\n", &vvdiv);
    return 0;
}
```

drodriguez commented:
> What was the motivation for this change?

The motivation is more or less replicated on the test. With the Xcode 16 SDKs there's a new "private" framework that does not exist in previous SDKs, but it was reexported from the "public" framework that existed in previous versions. The code in `isImplicitlyLinked` used to consider the libraries in `/usr/lib` and `/System/Library/Frameworks` linkable. The new "private" framework lives in one of those paths, so it was a candidate to be implicitly linked, with the bad luck that previous versions of the OS will not find the new file and fail to start the binaries. With the new code, libraries are only implicitly linked if the libraries do not have "allowable clients", which is the case in the new SDKs. Libraries with "allowable clients" should had never been linked directly, anyway.

> After this change, not only does lld not reject linking to vecLib.framework, it links to the library despite the binary not being an allowed client:

This was true *before* and *after* this change. LLD does not honor "allowable clients" at all. If you explicitly link against a private library, LLD will do what you ask. The changes above was trying to avoid a implicit link, but it was not trying to avoid any explicit link the user tries to do in the command line. I think nobody has taken the time to implement all the system of alloweable clients into LLD. It is of limited value to anybody that do not develop SDKs, which in the case of macOS, it is just Apple, and they use their own linker.

What this change was doing was that when linking to `Accelerate` (an allowable client of `vecLib`), `vecLib` will not be linked directly, but `Accelerate` will. Before this change, depending on the symbols, `vecLib` might have been linked incorrectly.

bdash commented:
Thanks for the explanation. Shortly before you posted your comment I found I had blamed this change in error and that, as you note, this change did not change the behavior I had observed. My apologies for that error.

> It is of limited value to anybody that do not develop SDKs, which in the case of macOS, it is just Apple, and they use their own linker.

There's a huge benefit to validating allowable clients for third-party developers that use Apple's SDKs. Apple uses the allowed clients list to limit what is considered part of the public ABI that they have to support. If you unintentionally link directly to a framework that restricts its allowable clients, your software will break in a software update if Apple moves symbols around. Apple uses their symbol versioning tricks to preserve binary compatibility when moving public symbols, but they're unlikely to consider that for libraries that they consider to be private. If the linker does not catch this class of mistake you can end up shipping a time bomb.

drodriguez commented:
My apologies. The phrase about being only useful to Apple was more about being able to link libraries restricted like that, but you are right that the restriction for linking against those libraries are still useful for everyone. We catch this problem early and avoided major issues, but it can be really bad if there's other "holes" in the implementation. Hopefully the issue you opened will catch the eye of someone.

carlocab commented:
Here's an implementation of restricting linking to libraries linked with `-allowable_client`: #114638

Please have a look if you have the time.
